src/memory/area_frame_allocator.rs

```diff
@@ -74,23 +80,72 @@ impl FrameAllocate for AreaFrameAllocator {
                 let address = area.base_addr + area.length - 1;
                 Frame::containing_address(address as usize)
             };
+            
+            let contained_by = |frame: &Frame, low: &Frame, high: &Frame| -> bool {
+               low <= frame && frame <= high 
+            };
 
-            if frame > current_area_last_frame {
-                // all frames of current area are used, switch to the new area
-                self.choose_next_area();
-            } else if frame >= self.kernel_start && frame <= self.kernel_end {
-                // 'frame' is used by the kernel
-                self.next_free_frame = Frame(self.kernel_end.0 + 1)
-            } else if frame >= self.multiboot_start && frame <= self.multiboot_end {
-                // 'frame' is used by the multiboot information structure
-                self.next_free_frame = Frame(self.multiboot_end.0 + 1)
-            } else {
-                // frame is unused, increment 'next_free_frame' and return it
+            let mut f_unused = true;
+            match frame {
```

Comment:
Note: `match` is an expression so
```rust
let bool;
match _ {
...
}
if bool {
```
could be
```rust
if match _ {
} {
```
Not that you have to change it, just an alternative.

Reply:
It might make it harder to read, especially if coming from another programming language... but "if match" makes it about 20-30 characters shorter, so ship it!